Source/DafnyRuntime/DafnyRuntime.cs

```diff
@@ -24,6 +24,86 @@ namespace Dafny {
   using System.Collections.Immutable;
   using System.Linq;
 
+  // Similar to Rune, which would be perfect to use
+  // except that it isn't available in the platforms we support
+  // (.NET Standard 2.0 and .NET Framework 4.5.2)
+  public readonly struct Rune : IComparable, IComparable<Rune>, IEquatable<Rune> {
+
+    private readonly uint _value;
+
+    public Rune(int value)
+      : this((uint)value) {
+    }
+
+    public Rune(uint value) {
+      if (!(value < 0xD800 || (0xE000 <= value && value < 0x11_0000))) {
+        throw new ArgumentException();
+      }
+
+      _value = value;
+    }
+
+    public int Value => (int)_value;
+
+    public bool Equals(Rune other) => this == other;
+
+    public override bool Equals(object obj) => (obj is Rune other) && Equals(other);
+
+    public override int GetHashCode() => Value;
+
+    // Values are always between 0 and 0x11_0000, so overflow isn't possible
+    public int CompareTo(Rune other) => this.Value - other.Value;
+
+    int IComparable.CompareTo(object obj) {
+      switch (obj) {
+        case null:
+          return 1; // non-null ("this") always sorts after null
+        case Rune other:
+          return CompareTo(other);
+        default:
+          throw new ArgumentException();
+      }
+    }
+
+    public static bool operator ==(Rune left, Rune right) => left._value == right._value;
+
+    public static bool operator !=(Rune left, Rune right) => left._value != right._value;
+
+    public static bool operator <(Rune left, Rune right) => left._value < right._value;
+
+    public static bool operator <=(Rune left, Rune right) => left._value <= right._value;
+
+    public static bool operator >(Rune left, Rune right) => left._value > right._value;
+
+    public static bool operator >=(Rune left, Rune right) => left._value >= right._value;
+
+    public static explicit operator Rune(int value) => new Rune(value);
+    public static explicit operator Rune(BigInteger value) => new Rune((uint)value);
+
+    public override string ToString() {
+      return char.ConvertFromUtf32(Value);
```

Comment:
Will this produce a Dafny-like literal? Will this `ToString()` method be used in a way that expects that?

Reply:
Ah good point, added this comment to clarify:

    // Defined this way to be consistent with System.Text.Rune,
    // but note that Dafny will use Helpers.ToString(rune) instead,
    // which will print in the style of a Dafny character literal.